bee-network/src/conns/mod.rs

```diff
@@ -0,0 +1,220 @@
+// Copyright 2020 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+mod dial;
+mod errors;
+mod manager;
+
+pub use dial::*;
+pub use errors::Error;
+pub use manager::*;
+
+use crate::{
+    interaction::events::{InternalEvent, InternalEventSender},
+    peers::{self, DataReceiver, PeerInfo},
+    protocols::gossip::{GossipProtocol, GossipSubstream},
+    PeerId, ShortId,
+};
+
+use futures::{prelude::*, select, AsyncRead, AsyncWrite};
+use libp2p::core::{
+    muxing::{event_from_ref_and_wrap, outbound_from_ref_and_wrap, StreamMuxerBox},
+    upgrade,
+};
+use log::*;
+use tokio::task::JoinHandle;
+
+use std::{fmt, sync::Arc};
+
+const MSG_BUFFER_SIZE: usize = 10000;
+
+pub(crate) async fn upgrade_connection(
+    peer_id: PeerId,
+    peer_info: PeerInfo,
+    muxer: StreamMuxerBox,
+    origin: Origin,
+    internal_event_sender: InternalEventSender,
+) -> Result<(), Error> {
+    let muxer = Arc::new(muxer);
+    let (message_sender, message_receiver) = peers::channel();
+
+    let internal_event_sender_clone = internal_event_sender.clone();
+
+    let substream = match origin {
+        Origin::Outbound => {
+            let outbound = outbound_from_ref_and_wrap(muxer)
+                .fuse()
```

Comment:
good catch. I think I forgot to remove it. Out of curiousity: Does it hurt to leave it in case the code gets refactored later on where it might be useful?

Reply:
Can you also have a look at the `Origin::Inbound` match, where there's an actual loop, but I am (for some reason) not fusing there. I probably wanted to fuse there, but I can't remember anymore :thinking:.